x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/scheduling/TransformScheduler.java

```diff
@@ -0,0 +1,236 @@
+     */
+    public interface Listener {
+        /**
+         * Called whenever scheduled task is being triggered.
+         *
+         * @param event event structure containing data that might be useful to the listener
+         */
+        void triggered(Event event);
+    }
+
+    private static final Logger logger = LogManager.getLogger(TransformScheduler.class);
+
+    private final Clock clock;
+    private final ThreadPool threadPool;
+    private final TimeValue schedulerFrequency;
+    private final TransformScheduledTaskQueue scheduledTasks;
+    /**
+     * Prevents two concurrent invocations of the "processScheduledTasks" method.
+     *
+     * Set to {@code true} before processing
+     * Set to {@code false} after processing (doesn't matter whether successful or not).
+     */
+    private final AtomicBoolean isProcessingActive;
+
+    public TransformScheduler(Clock clock, ThreadPool threadPool, Settings settings) {
+        this.clock = Objects.requireNonNull(clock);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.schedulerFrequency = Transform.SCHEDULER_FREQUENCY.get(settings);
+        this.scheduledTasks = new TransformScheduledTaskQueue();
+        this.isProcessingActive = new AtomicBoolean();
+    }
+
+    /**
+     * Starts the processing.
+     * Method {@code processScheduledTasks} is invoked according to its configured {@code schedulerFrequency}.
+     */
+    public void start() {
+        threadPool.scheduleWithFixedDelay(this::processScheduledTasks, schedulerFrequency, ThreadPool.Names.GENERIC);
+    }
+
+    // Visible for testing
+    void processScheduledTasks() {
+        // Prevent two concurrent invocations of the "processScheduledTasks" method
+        if (isProcessingActive.compareAndSet(false, true) == false) {
+            return;
+        }
+        logger.trace("Processing scheduled tasks started");
+        Instant processingStarted = clock.instant();
+        final boolean taskWasProcessed;
+        try {
+            taskWasProcessed = processScheduledTasksInternal();
+        } finally {
+            // Make sure we clear the "isProcessingActive" bit regardless of whether there was a success or failure.
+            // Otherwise, the processing would be stuck forever.
+            isProcessingActive.set(false);
+        }
+        Instant processingFinished = clock.instant();
+        logger.trace(
+            String.format(
+                Locale.ROOT,
+                "Processing scheduled tasks finished, took {}ms",
+                Duration.between(processingStarted, processingFinished).toMillis()
+            )
+        );
+        if (taskWasProcessed) {
+            // If we happened to process the task, there may be other tasks also eligible for processing.
+            // We try to process them ASAP as we don't want to wait the `delay` for every task.
+            processScheduledTasks();
```

Comment:
when a transform is triggered, it creates a task in the generic threadpool quickly returns. The task runs in _another_ thread and does the work, e.g. fires searches. My concern here: many transform can get triggered at once and fire searches nearly at the same time.

But this issue isn't new, it exists in the old scheduling. With this central scheduler we have the chance to fix this and e.g. add a jitter to avoid "thundering herd".

However, please understand this as "loud thinking", lets tackle that later.